### ISSUE TITLE
CRITICAL: Desktop - Chat Switcher is slow and doesn't recognize keystrokes

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#28071]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Type a message in a chat
2. Type `command+k` to open chat switcher
3. Immediately start typing the name of a person you already have chats with

## Expected Result:
The characters you type show up in search

## Actual Result:
 Many characters are missed. ie. I typed `lauren` quickly and only the `n` showed.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.73.0
**Reproducible in staging?:** n/a
**Reproducible in production?:** n/a
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

n/a

**Expensify/Expensify Issue URL:**
**Issue reported by:** @mallenexpensify 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1695286362554219

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011dba4ca10589e48c</li>
        <li>Upwork Job ID: 1707643423922569216</li>
        <li>Last Price Increase: 2023-10-17</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @mallenexpensify</details>

**Comment by stephanieelliott** [comment]
Adding relevant labels

**Comment by studentofcoding** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue

Desktop - Chat Switcher is slow and doesn't' recognize keystrokes 

## What is the root cause of that problem?

On [`src/pages/SearchPage.js`](https://github.com/Expensify/App/blob/dfb3f9b0b536d9640f81a85d59e095581dc06dba/src/pages/SearchPage.js#L58C16-L58C16) we have set a debounce of 75ms  `this.debouncedUpdateOptions = _.debounce(this.updateOptions.bind(this), 75);` This means that if the user types quickly, some keystrokes will be missed because the state update is debounced and does not happen immediately, when `onChangeText` function is called

## What changes do you think we should make in order to solve the problem?

We should move the call `this.debouncedUpdateOptions()` inside the callback function of setState(). This ensures that `debouncedUpdateOptions()` is only called after the state update is completed.

```
onChangeText(searchValue = '') {
    this.setState({searchValue}, () => {
        this.debouncedUpdateOptions();
    });
}
```

## Result

https://github.com/Expensify/App/assets/20473526/4c3c8400-d36f-4212-8f50-94f014e78ae1

## What alternative solutions did you explore? (Optional)

None

@cubuspl42 @stephanieelliott

**Comment by cubuspl42** [comment]
@studentofcoding 

> This means that if the user types quickly, some keystrokes will be missed because the state update is debounced and does not happen immediately, when onChangeText function is called

Could you elaborate? I understand why it does not happen immediately, but how does debouncing explain the missing keystrokes?

**Comment by studentofcoding** [comment]
> @studentofcoding
> 
> > This means that if the user types quickly, some keystrokes will be missed because the state update is debounced and does not happen immediately, when onChangeText function is called
> 
> Could you elaborate? I understand why it does not happen immediately, but how does debouncing explain the missing keystrokes?

Based on my observations the reason is because of race conditions when debouncedUpdateOptions() is called before the state is fully updated, thus causing some keystrokes to be missed @cubuspl42.

**Comment by cubuspl42** [comment]
@studentofcoding Can this race condition be observed when, for example, we add some `console.log` statements?

Could you provide a video with a sequence of logged events demonstrating the race? I'd like to understand better what kind of race condition we're dealing with. I believe that the solution might be working, but the root cause analysis is sometimes even more important than a fix itself.

**Comment by studentofcoding** [comment]
> @studentofcoding Can this race condition be observed when, for example, we add some `console.log` statements?
> 
> Could you provide a video with a sequence of logged events demonstrating the race? I'd like to understand better what kind of race condition we're dealing with. I believe that the solution might be working, but the root cause analysis is sometimes even more important than a fix itself.

Sure @cubuspl42, let me add the demonstration tonight

**Comment by stephanieelliott** [comment]
Hey @studentofcoding and progress with the demonstration?

**Comment by cubuspl42** [comment]
I [asked for help](https://expensify.slack.com/archives/C05LX9D6E07/p1696677854848959?thread_ts=1696418872.247919&cid=C05LX9D6E07) with figuring out this one on Slack

**Comment by mallenexpensify** [comment]
Posted again to try to get 👀 cuz I'd love to see this bug fixed, I run into it a dozen or more times a day. 
https://expensify.slack.com/archives/C05LX9D6E07/p1696870380595049?thread_ts=1696418872.247919&cid=C05LX9D6E07

**Comment by stephanieelliott** [comment]
Hey @studentofcoding, really interested to learn more about your theory around the race condition, can you share the video demonstrating the race? 

We'd really like to get this one fixed, so doubling the bounty on this issue to $1000.

**Comment by pradeepmdk** [comment]
when I observe before input focus typing character is missing which is expected only. but after focusing there are no issues here.
let me know if we are facing this issue after text focus.

**Comment by studentofcoding** [comment]
After another checking, The RCA I've shared only covers 30% of the problem, and the deeper root cause is found.

In that case, can I revise my proposal @stephanieelliott?

**Comment by cubuspl42** [comment]
@studentofcoding You can update your proposal, as stated in the contributing guidelines.

**Comment by studentofcoding** [comment]
Sure @cubuspl42, I've updated the [proposal](https://github.com/Expensify/App/issues/28071#issuecomment-1740849333)

cc: @stephanieelliott

**Comment by cubuspl42** [comment]
Well, the only issue is that you assume that we _can_ change a constant that was (probably experimentally?) decided to be 300 ms to 50 ms, without consequences. What's the rationale here?

This is the kind of an issue where hitting a nail in one place can _easily_ make things fall apart in another. We'll have to triple check that we're not doing exactly that.

**Comment by studentofcoding** [comment]
> Well, the only issue is that you assume that we _can_ change a constant that was (probably experimentally?) decided to be 300 ms to 50 ms, without consequences. What's the rationale here?
> 
> 
> 
> This is the kind of an issue where hitting a nail in one place can _easily_ make things fall apart in another. We'll have to triple check that we're not doing exactly that.

Yes, and that's why I shared note on the bottom of proposal that I'll make sure it's not making problem on another places (with the context that'll only related to Search Related input that get focused (after 300ms)-No data management or effect we do here)

**Comment by cubuspl42** [comment]
Your root cause analysis still doesn't feel convincing to me. For example, you say:

> This condition makes the **race condition**, that if the user types quickly (before the the Input isFocused) then some keystrokes will be missed.

> 1. Change **`CONST.ANIMATED_TRANSITION` to less than `this.debouncedUpdateOptions`** time (in the demo below I set ANIMATED_TRANSITION = 50)

For testing purposes, I manipulated the values in this way:

- `ANIMATED_TRANSITION = 50` (50 ms)
- `this.debouncedUpdateOptions = _.debounce(this.updateOptions.bind(this), 10)` (10 ms)

`ANIMATED_TRANSITION` is _bigger_ than the debounce duration, which doesn't seem to affect the solution.

I cannot see the relation between these two values, and I definitely can't see the race you're talking about.

**Comment by studentofcoding** [comment]
> Your root cause analysis still doesn't feel convincing to me. For example, you say:
> 
> > This condition makes the **race condition**, that if the user types quickly (before the the Input isFocused) then some keystrokes will be missed.
> 
> > 1. Change **`CONST.ANIMATED_TRANSITION` to less than `this.debouncedUpdateOptions`** time (in the demo below I set ANIMATED_TRANSITION = 50)
> 
> For testing purposes, I manipulated the values in this way:
> 
> * `ANIMATED_TRANSITION = 50` (50 ms)
> * `this.debouncedUpdateOptions = _.debounce(this.updateOptions.bind(this), 10)` (10 ms)
> 
> `ANIMATED_TRANSITION` is _bigger_ than the debounce duration, which doesn't seem to affect the solution.
> 
> I cannot see the relation between these two values, and I definitely can't see the race you're talking about.

I think you misunderstood my point:
- Based on RCA:
> `On [src/pages/SearchPage.js](https://github.com/Expensify/App/blob/dfb3f9b0b536d9640f81a85d59e095581dc06dba/src/pages/SearchPage.js#L58C16-L58C16) we have set a debounce of 75ms this.debouncedUpdateOptions = _.debounce(this.updateOptions.bind(this), 75); meanwhile we set focus to the Input ([BaseOptionsSelector.js](https://github.com/Expensify/App/blob/6f533d6730a1fa71a4b1d7529cd7dfc53aef8e29/src/components/OptionsSelector/BaseOptionsSelector.js#L74C9-L78C10)) after 300ms on [CONST.ANIMATED_TRANSITION.](https://github.com/Expensify/App/blob/6f533d6730a1fa71a4b1d7529cd7dfc53aef8e29/src/CONST.ts#L38C4-L38C30).

In simple terms, i**f we type before 300ms (or ANIMATED_TRANSITION is done), then the keystroke won't catch all the characters, because when the user types it's still not focused on the Search bar Input.**

- Also, we have a debouncedUpdateOptions of 75ms that will have a side effect of loading the data faster than what the user inputs on the Search bar (after 300ms)

**Therefore the flow right now gives a race condition with the order of** 
1. debouncedUpdateOptions, 
2. ANIMATED_TRANSITION, 
3. Data is rendered, 

**That Instead what we need is** 
1. ANIMATED_TRANSITION, 
2. debouncedUpdateOptions, and 
3. Data is rendered

That's basically the reason and flow of my solution @cubuspl42

**Comment by cubuspl42** [comment]
> I think you misunderstood my point

It's possible, but it's not the only explanation of our misunderstanding.

>  Also, we have a debouncedUpdateOptions of 75ms that will have a side effect of loading the data faster than what the user inputs on the Search bar (after 300ms)

> **Therefore the flow right now gives a race condition with the order of**
> 
>     1. debouncedUpdateOptions,
> 
>     2. ANIMATED_TRANSITION,
> 
>     3. Data is rendered,

How is it the case that `debouncedUpdateOptions` is called within these first 300 ms, when the input is not focused?

I thought that if the input is not focused, the characters can't be effectively typed-in, `onChangeText` can't be invoked, and in effect neither can be `debouncedUpdateOptions`. Could you show me the flow in my reasoning?

**Comment by cubuspl42** [comment]
The price tag has already been raised and should be enough for now.

What is expected from a good proposal here:
- A solid root cause analysis without logical flaws
- A link to the original PR introducing/changing the 300 ms constant
- Understanding _why_ it was introduced and decided to be 300 ms, not something else
  - Starting a dialogue with the author/C+ of the original proposal can be considered
- Proof of testing the proposed solution in the relevant area, showing that it doesn't add regression to the original problem

**Comment by cubuspl42** [comment]
I [advertised the issue on Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1697531886621219)

**Comment by cubuspl42** [comment]
It was suggested on Slack that the original issue worth investigating is https://github.com/Expensify/App/issues/6069

**Comment by AmjedNazzal** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Desktop - Chat Switcher is slow and doesn't recognize keystrokes

### What is the root cause of that problem?

First of all I should correct something about this bug report, this bug is NOT exclusive to desktop and I can easily reproduce it on web chrome using the same steps

https://github.com/Expensify/App/assets/74202751/f330e9d3-e1b4-48a3-9acd-ab038ed8baeb

We are using onChangeText to update the search value, this onChangeText is passed down all the way to RNTextInput which then animates the TextInput wrapping it in AnimatedTextInput, the issue here isn't really a bug but more of how it is intended to behave, so something like onChangeText is a react event and this kind of event along with an event like onKeyPress will not be attached until the component finishes mounting and in this case finishes animating and so anything typed before the animation is done isn't being captured by the native event.

### What changes do you think we should make in order to solve the problem?

If debouncing the input capture even further is not desirable what we can do is make use of searchRendered in searchPage, searchRendered is attached to onLayout and if we pass onLayout to RNTextInput through BaseOptionSelector and call it in BaseTextInput then that means that searchRendered will be called upon onLayout event which in TextInput native is an event that occurs right when the component mounts, from there we can add an event listener on the window object as soon as the component mounts:

```javascript
searchRendered() {
    window.addEventListener('keydown', this.handleKeyDown);
```

Now with that we have a listener for keydown events right when the TextInput component mounts regardless of the animation, from there we can capture any keydown events until react's own event handlers take over:

```javascript
handleKeyDown(event){
    this.setState(prevState => ({ 
        tempSearchValue: prevState.tempSearchValue + event.key
    }));
};
```
Once onChangeText event is attached it will override the keydown event effectivly make it as if it weren't there and keydown event we added will no longer work which is good, now we can combine both search values into one and hande over the control to onChangeText's native behaviour:

```javascript
onChangeText(searchValue = '') {
    ...
    
    if(this.state.tempSearchValue !== ''){
        this.setState({
            searchValue: this.state.tempSearchValue + searchValue,
            tempSearchValue: ''
        });
    }else {
        this.setState({searchValue}, this.debouncedUpdateOptions);
    }
```

lastly we would remove the event listener we added on componentWillUnmount

### Result

https://github.com/Expensify/App/assets/74202751/71629caa-81c8-43f0-a946-452019d89b8f

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by mallenexpensify** [comment]
Still reproducible for me, I just typed `cmd+k` then `Stephanie` and got this
<img width="321" alt="image" src="https://github.com/Expensify/App/assets/22508304/b029b74c-e420-4636-8ebd-6e426e52a1f1">

**Comment by cubuspl42** [comment]
@AmjedNazzal Do I understand correctly that you disagree that this issue is related to the `CONST.ANIMATED_TRANSITION` constant?

**Comment by AmjedNazzal** [comment]
@cubuspl42 No I don't disagree, that duration of animation is related to my proposal, what I was trying to propose is that if manipulating the animation duration and using a debounce approach is not desirable then I was proposing an alternative solution where the animation would remain the same, again in my proposal I explain that until animations are done, react's native events won't be attached so in my opinion there is no straight forward solution as far as I can see and we could either manipulate the animation duration or we could attach a window event listener as a placeholder until react's event listeners are attached.

**Comment by shubham1206agra** [comment]
@cubuspl42 This is more of a performance improvement. We should not modify the `CONST.ANIMATED_TRANSITION` as this will break animation of screen switching.

**Comment by cubuspl42** [comment]
@shubham1206agra 

> @cubuspl42 This is more of a performance improvement. We should not modify the `CONST.ANIMATED_TRANSITION` as this will break animation of screen switching.

Are you sure about this statement? My conclusion so far is that nothing (except maybe the title) suggests that this actually is a performance problem, even I thought that it has a performance-related aspect in the beginning.

**Comment by shubham1206agra** [comment]
Yes
Because I tested this issue on 2 different models
And the number of keystrokes missing are different

If someone gets the bug then I may be wrong but I am pretty confident that this is due to performance

**Comment by shubham1206agra** [comment]
> @shubham1206agra 
> 
> > @cubuspl42 This is more of a performance improvement. We should not modify the `CONST.ANIMATED_TRANSITION` as this will break animation of screen switching.
> 
> Are you sure about this statement? My conclusion so far is that nothing (except maybe the title) suggests that this actually is a performance problem, even I thought that it has a performance-related aspect in the beginning.

You can't reduce the transition time as it may introduce some new problems

**Comment by cubuspl42** [comment]
@shubham1206agra 

Reducing `ANIMATED_TRANSITION` to 75 ms makes the problem disappear on my laptop. Why would it happen if this was a performance problem? Also, if this is a performance problem, than of what nature?

> You can't reduce the transition time as it may introduce some new problems

Well, we definitely _can_ do that, but I'm not sure yet whether we should. If you actually know about some part of the app that could likely break after reducing the mentioned constant, please let us know, so we can test it!

**Comment by shubham1206agra** [comment]
> @shubham1206agra 
> 
> Reducing `ANIMATED_TRANSITION` to 75 ms makes the problem disappear on my laptop. Why would it happen if this was a performance problem? Also, if this is a performance problem, then of what nature?
> 

It's due to focus is called a little early.

Not sure of the nature of the performance improvement. But this may be React using some batch updates and focus is taking longer than `ANIMATED_TRANSITION` time (not sure)

> > You can't reduce the transition time as it may introduce some new problems
> 
> Well, we definitely _can_ do that, but I'm not sure yet whether we should. If you actually know about some part of the app that could likely break after reducing the mentioned constant, please let us know, so we can test it!

Wherever we have used setTimeout without using const right now (Which is wrong). I have to double check
And maybe timing was decided by the UI team. You may want to consult them before changing time

**Comment by getusha** [comment]
IMO reducing the `ANIMATED_TRANSITION` is definitely a hack and it is not fixing the root cause, additionally the issue will still be either reproducible even without the setTimeout or will cause a glitch.

**Comment by cubuspl42** [comment]
So far, no proposal seems to be acceptable

@studentofcoding You didn't reply to [my questions](https://github.com/Expensify/App/issues/28071#issuecomment-1761315845)

@AmjedNazzal You seem to have focused on making your proposal different from the other, while we are seeking for a solid proposal with the best solution (or at least reasonably good...) without logical flaws

> If debouncing the input capture even further is not desirable...

Before reasoning whether it's desirable, can we first determine whether it does or does not solve the problem? Because just increasing the debounce window like this...

```js
this.debouncedUpdateOptions = _.debounce(this.updateOptions.bind(this), 375);
```

...doesn't seem to solve the problem, as I observe it.

**Comment by cubuspl42** [comment]
@getusha 

> IMO reducing the `ANIMATED_TRANSITION` is definitely a hack

Do you have any arguments for this claim?

I don't have strong feelings about the `ANIMATED_TRANSITION` constant, but in general, if the transition is long as it is for esthetic reasons, and it actually affects UX of a user who wants to make things done while typing _reasonably_ fast, then it's **not impossible** that the transition is just too long.

> additionally the issue will still be either reproducible even without the setTimeout

I'm not sure if I got this right. Are you saying that reducing `ANIMATED_TRANSITION` doesn't fix the issue in your case, i.e. you're still reproducing it after reducing `ANIMATED_TRANSITION`?

>  or will cause a glitch.

It's possible. What kind of glitches do you have in mind, or is there any specific part of the app that you would test first? Or, alternatively, do you have any past issue in mind that might return after we reduce the mentioned constant?

**Comment by cubuspl42** [comment]
We don't have an infinite set of options here, I want to fix this issue, as closing it doesn't feel appropriate; I agree that there is a problem here. I encourage a constructive approach based on evidence or experience. Also, proposals without logical flaws.

**Comment by cubuspl42** [comment]
Also, everyone, please note that I suggested [an investigation path](https://github.com/Expensify/App/issues/28071#issuecomment-1765952285) that nobody picked up, which still might be fruitful.

> * A link to the original PR introducing/changing the 300 ms constant
> 
> * Understanding _why_ it was introduced and decided to be 300 ms, not something else
>   
>   * Starting a dialogue with the author/C+ of the original proposal can be considered
> 
> * Proof of testing the proposed solution in the relevant area, showing that it doesn't add regression to the original problem

**Comment by cubuspl42** [comment]
It appears that there is [a Slack thread](https://expensify.slack.com/archives/C05LX9D6E07/p1697639189667039) on the internal performance-related channel. We'll have to coordinate efforts!

**Comment by mallenexpensify** [comment]
@cubuspl42 thanks for the link! Good sleuthing, I hadn't seen.

**Comment by mallenexpensify** [comment]
I'm bumping this to weekly for now, since there's a discussion going on.

**Comment by stephanieelliott** [comment]
Coming from this [Slack summary ](https://expensify.slack.com/archives/C05LX9D6E07/p1698193581128549) it seems like this PR may fix the issue https://github.com/Expensify/App/issues/30261

**Comment by cubuspl42** [comment]
I'm not sure why https://github.com/Expensify/App/issues/30261 was created, it sounds like a dupe of this one. I asked a question there

**Comment by stephanieelliott** [comment]
https://github.com/Expensify/App/issues/30261 still being worked on

**Comment by cubuspl42** [comment]
Let's hold on https://github.com/Expensify/App/issues/30261

**Comment by stephanieelliott** [comment]
Added HOLD status to title

**Comment by cubuspl42** [comment]
https://github.com/Expensify/App/pull/30263 has been merged; we can put off hold and re-test

**Comment by stephanieelliott** [comment]
Hey @kbecciv can you retest this one please?

**Comment by kbecciv** [comment]
@stephanieelliott Checking, will update you shortly!

**Comment by kbecciv** [comment]
Reproducible with v1.3.96-6.
When typing "Dwight" but only "wight" is entered in the search field.

https://github.com/Expensify/App/assets/93399543/7b052c6a-78f2-4ca0-bc72-9d53827085e4

**Comment by mallenexpensify** [comment]
Still happening for me, but it seems to be better after the below was implemented, just commented there
- https://github.com/Expensify/App/issues/30261#issuecomment-1802733010

**Comment by cubuspl42** [comment]
Summing up: we know that reducing the `ANIMATED_TRANSITION` used to help with our bug. After the optimization (https://github.com/Expensify/App/issues/30261), maybe reducing it _less_ will make things right here?

At the same time, I'm aware that even touching `ANIMATED_TRANSITION` is somewhat risky, as we rely on it in many places. This makes our issue tricky.

**Comment by mallenexpensify** [comment]
@cubuspl42 do you think this bug is possibly related? ie. that some version of animation is making it take longer to load and recognize keystrokes? https://expensify.slack.com/archives/C049HHMV9SM/p1699896354633069

**Comment by cubuspl42** [comment]
@mallenexpensify

In a way, for sure. This issue has two components: the delay introduced by `ANIMATED_TRANSITION`, and the performance-related delays. The linked case nearly for sure has at least one of these components.

**Comment by mallenexpensify** [comment]
I find it a bit frustrating that we're trading off style and animation over usability.  Let's keep the issues separate for now, Im' curious what proposals we might get for the other issue.

**Comment by cubuspl42** [comment]
It's not done on purpose, I think.

- Nobody said it's unacceptable to lower `ANIMATED_TRANSITION`, it's just a bit "scary", as it's a global app-wise constant. Something like "move this single thing here and shake the whole app".
- We're also dealing with a platform quirk, specifically on the Web; as far as I understand, it's not possible to focus the inputs until the animation is 100.0% complete
- We're dealing with animation **and** performance issues at the same time, it doesn't help

**Comment by mallenexpensify** [comment]
Thanks @cubuspl42 , I was mostly just whining/griping there for a min.  Do you think it's best to put this on hold for now, move to monthly and remove `Help Wanted`?  I'd remove Steph and assign myself.  I'm hesitant to _just_ close the issue cuz I'd love to see it fixed at some point, even if it's many months down the road.

**Comment by cubuspl42** [comment]
@mallenexpensify

Honestly, I'm not optimistic about this being fixed externally...

From the engineering point of view, the next steps could be:

- Investigating how animations and computation performance contribute to the observable issue _after_ https://github.com/Expensify/App/issues/30261#issuecomment-1802733010 was merged
- Creating an objective, automated end-to-end performance test that would allow us to measure the impact of potential solutions
- Investigating whether the `ANIMATED_TRANSITION` constant could be lowered, having in mind the app-wise consequences

> put this on hold for now, 

I'm not sure what we would hold on.

> move to monthly and remove Help Wanted?

Then we'd need to assign this to an expert agency or an internal engineer, right?

**Comment by mountiny** [comment]
@mallenexpensify I think we should close this issue, this is unfortunately basically impossible to fix with the current set up and while its annoying we just need to move on, we are going to update the search soon as part of the ideal-nav so maybe we can make it better there

**Comment by mallenexpensify** [comment]
I'm going to assign to myself, remove others and make it a monthly. Then, drop into an ideal nav room when the time is right, to insure it's being considered.  Thx @mountiny and @cubuspl42

**Comment by mallenexpensify** [comment]
Checking in Wave 9, that used to be Ideal Nav
https://expensify.slack.com/archives/C05NJ4SLBMF/p1703259924998969

**Comment by mallenexpensify** [comment]
Checking link above to see where this belongs

**Comment by mallenexpensify** [comment]
Sorry @0xmiroslav , was testing auto-assigner automation

**Comment by mallenexpensify** [comment]
This appears to have gotten worse.  And I fear it's more likely to get worse than better unless we do something about. A minute ago I 
- Opened chat switcher
- Waited 2-3 seconds, knowing it's not recognizing keystrokes
- Typed `social` and then `return` (pretty quickly)
- Chat opened to `Shawn` somehow

**Comment by filip-solecki** [comment]
Hi! I am Filip Solecki from Software Mansion - an expert agency and I'd like to work on this issue.

**Comment by mallenexpensify** [comment]
Thanks @filip-solecki , one other issue that _might_ be related is, it's possible for me to type someone's name and hit enter (quickly), but then the correct chat doesn't open.  I can easily reproduce if you want a vid, can also put it on hold and address after the fix for keystrokes not being recognized when first typed.

**Comment by filip-solecki** [comment]
@mallenexpensify Please upload video mentioned above.

**Comment by mallenexpensify** [comment]
Per discussion [here](https://expensify.slack.com/archives/C04878MDF34/p1705776345530929?thread_ts=1705614086.221599&cid=C04878MDF34)
I've put this on hold pending the Fabric update.  Updated to weekly and I'll do my best to be patient every... single... time I have to wait a second or two before I start typing after typing `cmd+k`  :)

**Comment by filip-solecki** [comment]
I think I requested for high-traffic account but you can check: filip.solecki@swmansion.com.

**Comment by mallenexpensify** [comment]
I requested to have you added to a high traffic account @filip-solecki

**Comment by mallenexpensify** [comment]
Here's a 'real world' example of why this is so frustrating.  I think it might be getting, or have gotten, worse too. I've had a dozen instances this morn of the bug. 
- cmd+k
- start typing `Faye`
- Assume I'm in my chat with Faye, send three messages to Faye
- Notice I'm actually in my chat with Joanie for some reason. 

https://github.com/Expensify/App/assets/22508304/5321a35c-8b44-4803-8f05-032e8838cf8e

**Comment by mallenexpensify** [comment]
Sasha mentions search is snappier [here](https://expensify.slack.com/archives/C05LX9D6E07/p1707349905667519). Has anything been done that'd help explain that?  I just did some testing and I agree that it appears to be better, the characters I type don't go missing like they did before.

**Comment by mallenexpensify** [comment]
Different, but related, bug.  Keystrokes are being recognized but, after entering a few characters and clicking enter, the _correct_ chat doesn't load, the old/existing chat persists

https://github.com/Expensify/App/assets/22508304/6b348eda-481a-4fd0-83f5-9999c978e8c7

**Comment by mallenexpensify** [comment]
Still occasionally run into issues so leaving this open a lil while longer, will add more examples soon.

**Comment by mallenexpensify** [comment]
For this example, this is my _normal_ typing speed and I can reproduce the _bug_ multiple times. 

https://github.com/Expensify/App/assets/22508304/ac78c5e9-c7b5-4deb-b0e4-d61bb84e6722

**Comment by mallenexpensify** [comment]
Holding on
- https://github.com/Expensify/App/issues/8503

**Comment by mallenexpensify** [comment]
Still waiting on Fabric update which should be soon 🤞 
https://github.com/Expensify/App/issues/8503

(posting cuz I just started my day and instantly sent a message to the wrong person cuz of the lag in search)

**Comment by mallenexpensify** [comment]
Unsure what's best here.  Between chat switcher and the compose box, there seems to be multiple issues with being slow or laggy :/

**Comment by mallenexpensify** [comment]
Tested again on latest staging version, still happening, regularly missing 1 or 2 keystrokes
<img width="237" alt="image" src="https://github.com/Expensify/App/assets/22508304/18280cc8-1368-4e68-bf35-627cca89dafe">

<img width="393" alt="image" src="https://github.com/Expensify/App/assets/22508304/7b498ea8-13de-4978-af68-c6c1265a8a7f">

**Comment by mallenexpensify** [comment]
Still a lil slow, I typed `lucien` here then pressed enter quickly


https://github.com/Expensify/App/assets/22508304/074f163f-58ea-47d2-b9de-a8dd77023223

**Comment by mallenexpensify** [comment]
We're off hold, Fabric is live and the chat switch is still a lil slow, at least for me. In a few tests I did right now, it was just the first letter that doesn't show, if you type quickly after typing `cmd+k` in Desktop. @filip-solecki , any thoughts on next best steps here?

**Comment by filip-solecki** [comment]
I will try to find the root cause and hopefully also some solution in a few days

**Comment by filip-solecki** [comment]
Hi! Coming back after investigation. 
The problem is in the BaseSelectionList. In the screenshot I uploaded, you can see that we deliberately delay the focus on the input. This was added a long time ago to fix the focus problem when going back to the page with SelectionList. Basically, the only roving we can do in this case is to conditionally bypass setTimeout, in this particular case. It seems that just in this case the error mentioned above should not occur, because from the ChatSwitcher page it is not possible to go anywhere further and then go back to it.

<img width="807" alt="Screenshot 2024-05-17 at 11 01 13" src="https://github.com/Expensify/App/assets/153545991/2eda31fe-646a-4369-8d32-26347a436aec">

@mallenexpensify

**Comment by mallenexpensify** [comment]
Thanks @filip-solecki .  Does this affect other parts of the app too? I've been noticing that the first character often gets missed.  ie. here I typed `black` and only `lack` shows. 
<img width="370" alt="image" src="https://github.com/Expensify/App/assets/22508304/4d3659e9-8ae3-4639-9c47-5f7bd5ef01cc">

- https://github.com/Expensify/App/issues/25893
@thiagobrez , @roryabraham , since you worked on the above.  Can you review the above and provide more context/feedback? Nearly every time I `cmd+k` to open the chat switcher and start typing, the first character (or more) don't show.  I'd love to fix this if we can.

**Comment by filip-solecki** [comment]
Currently it can occur on every page with SelectionList when you start typing before animation transition is over.

**Comment by mallenexpensify** [comment]
Thanks @filip-solecki .  What do you propose are the next best steps here?  (I'm unfamiliar with SelectionList and def not an engineer).

**Comment by filip-solecki** [comment]
The simplest solution might be to provide an option to bypass the timeout causing missing letters on certain pages where it's possible. This is not a perfect solution since there are pages that need this timeout to prevent previously mentioned bugswhen navigating back.

Another option is to discover an alternate solution for the aforementioned bug. The current timeout is essentially a workaround. If we can find a different approach, we won't need the timeout and the related issue will be eliminated.

**Comment by mallenexpensify** [comment]
Thanks @filip-solecki. Since this is a part of #newdot-quality and you're a member of the channel, can you start a conversation there about how we want to proceed?  With more 👀, we might be able to find a solution that covers all bases. Thx

**Comment by kevinksullivan** [comment]
followed up in slack

https://expensify.slack.com/archives/C05LX9D6E07/p1717540088086959?thread_ts=1716804890.419359&cid=C05LX9D6E07

It seems like we either need to find someone internal to get on this and make it `Daily` if we think it's truly CRITICAL, or lower the priority level so it is not CRITICAL.

**Comment by mallenexpensify** [comment]
PR is in review.  Let me know if I can hellp with testing, I have a feeling regressions on this could be pretty bad.

**Comment by mallenexpensify** [comment]
@aimane-chnaif can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01bcbf7accb280d408

I reviewed the issue and PR, there are pretty much no comments on this issue but you reviewed the PR and made a couple comments, so I set the price at $250.

**Comment by mallenexpensify** [comment]
Contributor+: @aimane-chnaif paid $250 via Upwork. 

@aimane-chnaif , for urgent issues,[ they need to have the `High Priority` label in order to be auto-increased in price](https://expensify.slack.com/archives/C01GTK53T8Q/p1717451725980889?thread_ts=1712256409.152489&cid=C01GTK53T8Q). The issue you linked to was an error that I'm addressing internally.
